backend: gl: add dither

yshui opened the pull request:
Add bayer ordered dithering when presenting to screen. Reduce banding
when using a strong blur. Also use 16-bit intermediary textures to
preserve precision in the rendering pipeline.

Related: #602

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com><!-- Please enable "Allow edits from maintainers" so we can make necessary changes to your PR -->

yshui committed:
backend: gl: add dither

Add bayer ordered dithering when presenting to screen. Reduce banding
when using a strong blur. Also use 16-bit intermediary textures to
preserve precision in the rendering pipeline.

Related: #602

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
options: add dithered-present option

See also 0a2cd0f14eebc41cc4c7b7c9a3db964a2b1a5ab0

Related: #602

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
backend: gl: don't add dither where it's not needed

If a pixel is perfectly representable as an 8-bit number, don't add
dither. Reduce artifacts where dither is unnecessary.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
backend: gl: try different back buffer formats

Prefer RGB formats first, because they use less memory; but fallback to
RGBA formats, as they are formats required by OpenGL.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

tryone144 commented on src/backend/gl/gl_common.c:
```diff
@@ -907,14 +916,25 @@ bool gl_init(struct gl_data *gd, session_t *ps) {
-	                       gd->back_texture, 0);
 	glDrawBuffer(GL_COLOR_ATTACHMENT0);
-	if (!gl_check_fb_complete(GL_FRAMEBUFFER)) {
+	const GLint *format = (const GLint[]){GL_RGB8, GL_RGBA8};
+	if (gd->dithered_present) {
+		format = (const GLint[]){GL_RGB16, GL_RGBA16};
+	}
```
ASAN does not like how format is initialized and then used below for `gd->back_format`. Fails with *stack-use-after-scope* in line 927. (only when run with `--dithered-present`)
This is most likely a false-positive as the referenced value is an integer (`typedef int GLint;`) and used by value not by reference...

```suggestion
	const GLint *format = gd->dithered_present ? (const GLint[]){GL_RGB16, GL_RGBA16}
	                                           : (const GLint[]){GL_RGB8, GL_RGBA8};
```

yshui replied:
ah good catch!

yshui replied:
yeah technically this is illegal as the temporary went out of scope. in practice `(const GLint[])` is probably stored in `.rodata`. but i should fix this.

tryone144 reviewed (approved):
Nice work! :+1:
Using 16bit textures requires ~25MB more VRAM than 8bit ones after a short test on my nvidia card.

Are you sure you wanted to add the README changes into this PR? ;)

---
I'm a bit behind with my reviews on the previous PRs. Sorry for that. The changes to the masking and shadow for shaped windows were rather large and consecutive time is rather limited at the moment. :disappointed:

yshui commented:
> add the README changes into this PR? ;)

oh no, i meant to push that to `next` :smiling_face_with_tear:

yshui committed:
backend: gl: fix use-after-scope

'format' was pointing to an array with a shorter lifetime

suggested by @tryone144

Co-authored-by: Bernd Busse <bernd@busse-net.de>

yshui commented:
> I'm a bit behind with my reviews on the previous PRs. Sorry for that. 

Just want to say I appreciate your help! :+1:  and take your time, there are long period of time where I can't put any time on this project too, so I can understand. (also the reason why i think this project needs more than just two of us).
